### PR TITLE
IOS-1967: Handle mime type check in decodable requests

### PR DESCRIPTION
**Pull request by rhogan11**
**Issue Link**
https://spothero.atlassian.net/browse/IOS-1967

**Description**
This is marked with proposal because I just wanted to get some feedback on the solutions here.  If we are simply checking the content type is `application/json` in the response handler, I think just uncommenting the existing code is fine for the non-combine request builders.  Are we trying to do more here, like match content types against any Accept headers if they are there?

Also, for implementing this on the combine version, I think we have a couple options.  
1) expose the full `URLSession.DataTaskPublisher.Output` down the chain
2) pass in some kind of validation to the base call site.  I wrote up a very quick example of what this could look like.

option 1 feels like better use of combine, but I think option 2 could works as well.

@bdrelling I was wondering a) why you originally commented out the guard statement? Just uncommenting it out makes me feel like I am missing something and b) what are your opinions on implementing a similar check in the combine version?

### CODE REVIEW

**Comment by rhogan11** [comment]
test are failing because we are not stubbing with the right content type, but ill hold off on fixing that for now

**Review by bdrelling — APPROVED** [review]

**Comment by bdrelling** [comment]
Copied from Slack for convenience:

> kk so two quick things I got from a minute or two of re-familiarizing myself:
>
> 1. Check out `StubResponse` for how it currently encapsulates the data of a response. I’d probably just add something like `MimeType?` there as a property, once you have the `MimeType` struct like above, or if you decide to just make a constant, a `String`. Then, you’ll want to update `MockURLProtocol` to set `response.mimeType` to `stubResponse.mimeType?.rawValue`, And should be good to go. Once all this is set up, you’ll just need to make sure that any time you create a `StubResponse` that is encodable, you automatically set this value. I’d do it in `StubResponse` convenience extensions at the bottom of the file.
> 2. Given the ease of extending `StubResponse`, should we extend `StubRequest` to additional allow validators and/or mimeType and so on that it can validate against? This may sound useless in terms of matching our stubs, BUT keep in mind how our `MockURLProtocol` works to search for stubbed responses and resolve conflicts if there’s multiple matches.

File: Sources/UtilityBeltNetworking/Models/MimeType.swift [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// Copyright © 2021 SpotHero, Inc. All rights reserved.
+
+public enum MimeType: String, Codable {
+    case json = "application/json"
+}
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/Models/MimeType.swift:5]: I'm still thinking this should be a struct so that `ensureMimeType` can be expanded upon without necessarily requiring updates within the `UtilityBelt` layer.

For example, I use UB for a lot of non-SH projects as a way of dogfooding apps with this library, and there are one or two cases I've needed custom mime types that a library might not (and shouldn't) account for by default. This is similar to why the Headers dictionary is keyed with `Strings` under the hood--there are custom headers for various use cases and we need to allow someone to override it.

Basically my own rule of thumb is: if something is _not_ a definitive, finite list (not implying it can't grow, but that we know all the possible values _at this time_ and don't see an immediate need for other values), it needs to _not_ be an `enum` type or we back ourselves into a corner. Does that make sense?

File: Sources/UtilityBeltNetworking/HTTPClient.swift [hunk 2]
```diff
@@ -208,11 +208,10 @@ public class HTTPClient {
...
                     }
                 case let .success(data):
-                    // TODO: Implement mime type checking for JSON before attempting to decode JSON (IOS-1967)
-//                    // If the mime type for the response isn't JSON, we can't decode it
-//                    guard dataResponse.response?.mimeType == "application/json" else {
-//                        return .failure(UBNetworkError.invalidContentType(dataResponse.response?.mimeType ?? "unknown"))
-//                    }
+                    // If the mime type for the response isn't JSON, we can't decode it
+                    guard dataResponse.response?.mimeType == MimeType.json.rawValue else {
+                        return .failure(UBNetworkError.invalidContentType(dataResponse.response?.mimeType ?? "unknown"))
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient.swift:213]: Why not re-use your validators for this? It doesn't make this call site 

```suggestion
                    // If the mime type for the response isn't JSON, we can't decode it
                    let validator = Validators.ensureMimeType(.json)
                    
                    do {
                        try validator(dataResponse.response?.mimeType)
                    } catch {
                        return .failure(error)
                    }
```

File: Sources/UtilityBeltNetworking/HTTPClient+Combine.swift [hunk 3]
```diff
@@ -7,20 +7,39 @@
...
+            /// - Returns: The validator that will do this check.
+            static func ensureMimeType(_ mimeType: MimeType) -> Validator {
+                return { element in
+                    let responseMimeType = element.response.mimeType
+                    if responseMimeType != mimeType.rawValue {
+                        throw UBNetworkError.invalidContentType(responseMimeType ?? "unknown")
+                    }
+                }
+            }
+        }
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:26]: I wouldn't mind having `Validators` in its own file! Especially given the other suggestion of reusing this across both clients.

File: Sources/UtilityBeltNetworking/HTTPClient.swift (hunk 2 above)
Comment (bdrelling) [reply Sources/UtilityBeltNetworking/HTTPClient.swift:213]: In fact, thinking about its ore, I think you could even use the same loop for validators and everything. Any reason not to implement it the same way?

**Review by bdrelling — CHANGES_REQUESTED** [review]
Really good stuff so far and headed down the right path! A few suggestions that might result in a bit more work.

Let me know if you want any help or pairing on some of this stuff. I _think_ everything mentioned is within reasonable scope of work with this PR, but I'm happy to help as I know this is runoff from Product Health Day, so just let me know!

File: Sources/UtilityBeltNetworking/HTTPClient.swift (hunk 2 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/HTTPClient.swift:213]: ya I like that

File: Sources/UtilityBeltNetworking/Models/MimeType.swift [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright © 2021 SpotHero, Inc. All rights reserved.
+
+import Foundation
+
+public struct MimeType: RawRepresentable, Codable {
```
Comment (rhogan11) [inline Sources/UtilityBeltNetworking/Models/MimeType.swift:5]: @bdrelling this is the only way I found to make a struct act like a string enum, such that a user of `UtilityBeltNetworking` could add extra "cases" in an extension.  Doesn't seem right tho.... Is there a different protocol to use here that would accomplish this better?

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            let responseMimeType = response.mimeType
+            if responseMimeType != mimeType.rawValue {
+                throw UBNetworkError.invalidContentType(responseMimeType ?? "unknown")
+            }
+        }
+    }
+    
+    /// Creates a validator that will ensure the response status code is not an error code.
+    /// - Returns: The validator that will do this check.
+    static let validateStatusCode = ResponseValidator { response in
```
Comment (rhogan11) [inline Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:38]: I added this validator as well given that we are using the validator pattern in both the combine/non-combine request builders.  I'm not sure I see where this check was being done on the non-combine version?  To be clear, we are adding it in this PR now.

**Comment by rhogan11** [comment]
@bdrelling just a heads up this is ready for review

File: Sources/UtilityBeltNetworking/HTTPClient+Combine.swift [hunk 6]
```diff
@@ -110,13 +110,20 @@
                                             parameters: ParameterDictionaryConvertible? = nil,
                                             headers: HTTPHeaderDictionaryConvertible? = nil,
                                             encoding: ParameterEncoding? = nil,
+                                            validators: [ResponseValidator] = [],
                                             dispatchQueue: DispatchQueue = .main,
                                             decoder: JSONDecoder = JSONDecoder()) -> AnyPublisher<T, Error> {
+            // Given that we expect to decode JSON from the response, add
+            // an additional validator that checks the appropriate mime type.
+            var validators = validators
+            validators.append(.ensureMimeType(.json))
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:119]: What happens if someone already added this validator, not knowing it was included by default?

I see three possible ways to avoid conflicts:

1. Make the signature `validators: [ResponseValidator] = [.ensureMimeType(.json)]`. This is a bit weird for the `Decodable` version of these requests, because we _know_ someone wouldn't want to drop them. But it would be clear they are being overridden, and it would allow someone to override in case they had another deserialization format that didn't use `JSON`. (This _is_ possible to have, such as with Firebase Firestore or GRPC, which both have `Codable` compatibility.)
2. Make the array a `Set` type. You might need to add some protocols to the `ResponseValidator` type, but should be fairly simple.
3. Check to see if the validator exists, and don't add it if so. I think to equate this you'll need to implement the same protocols as you would for #2.

File: Sources/UtilityBeltNetworking/HTTPClient.swift [hunk 7]
```diff
@@ -30,6 +30,11 @@ public class HTTPClient {
     /// set to be more restrictive. https://stackoverflow.com/a/54806389
     public var timeoutInterval: TimeInterval?
     
+    /// Validators that will be applied to all responses.
+    private let defaultValidators: [ResponseValidator] = [
+        .validateStatusCode,
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient.swift:35]: Can someone override this to _not_ validate the status code?

File: Sources/UtilityBeltNetworking/HTTPClient.swift [hunk 8]
```diff
@@ -186,15 +204,22 @@ public class HTTPClient {
...
                                       headers: HTTPHeaderDictionaryConvertible? = nil,
                                       encoding: ParameterEncoding? = nil,
+                                      validators: [ResponseValidator] = [],
                                       dispatchQueue: DispatchQueue = .main,
                                       decoder: JSONDecoder = JSONDecoder(),
                                       completion: DecodableTaskCompletion<T>? = nil) -> URLSessionTask? {
+        // Given that we expect to decode JSON from the response, add
+        // an additional validator that checks the appropriate mime type.
+        var validators = validators
+        validators.append(.ensureMimeType(.json))
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient.swift:214]: Same as above re: conflicts and incorporating one of the options.

File: Sources/UtilityBeltNetworking/HTTPClient.swift [hunk 9]
```diff
@@ -306,6 +328,17 @@ public class HTTPClient {
         
         print("[UtilityBeltNetworking] \(message)")
     }
+    
+    /// Validates a response against `defaultValidators` and passed in validators.
+    /// - Parameters:
+    ///   - response: The response to perform the validation on.
+    ///   - validators: Additional validators on top of the default validators that should be applied to the response.
+    /// - Throws: Throws any validation errors, indicating failed validation.
+    func validate(response: HTTPURLResponse, with validators: [ResponseValidator]) throws {
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/HTTPClient.swift:337]: How can someone using this client opt out of this validation if they need to?

File: Sources/UtilityBeltNetworking/Models/MimeType.swift (hunk 4 above)
Comment (bdrelling) [reply Sources/UtilityBeltNetworking/Models/MimeType.swift:5]: No, this looks totally great! What do you think doesn't seem right about it? I assumed the protocol was `RawStringConvertible` or something but this actually seems more appropriate.

The only way you could make it simpler is by incorporating `ExpressibleByStringLiteral`, but I don't mind how you have it right now honestly.

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    func validate(response: HTTPURLResponse) throws {
+        try self.validationCheck(response)
+    }
+}
+
+// MARK: - Common Validators
+
+public extension ResponseValidator {
+    /// Creates a validator that will ensure the response has a specific mime type
+    /// - Parameter mimeType: The mime type we want to validate the existance of.
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:25]: Typo

```suggestion
    /// - Parameter mimeType: The mime type we want to validate the existence of.
```

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift (hunk 5 above)
Comment (bdrelling) [reply Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:38]: If I'm understanding and remembering correctly, we did not validate status codes before, correct?

If not, this may be a breaking change requiring some testing in the app. I think in some ways we rely on a failed status code to then include an error message and some data, so while including this sort of validator is great, I don't know that we should be including this in any defaults.

What are your thoughts?

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        }
+    }
+    
+    /// Creates a validator that will ensure the response status code is not an error code.
+    /// - Returns: The validator that will do this check.
+    static let validateStatusCode = ResponseValidator { response in
+        let status = response.status
+        if status.responseType == .clientError || status.responseType == .serverError {
+            // TODO: Update with better error
+            throw UBNetworkError.unexpectedError
```
Comment (bdrelling) [inline Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:42]: Are you planning to cover this in this PR, just `TODO` per your above message?

It's fairly straight-forward, so would definitely be good to include, especially considering potential impact if we leave this as a newly introduced default check.

**Review by bdrelling — CHANGES_REQUESTED** [review]

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift (hunk 11 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:42]: I kept the same status code checking that was previously in the combine HTTPClient, and that code included this TODO

File: Sources/UtilityBeltNetworking/Models/MimeType.swift (hunk 4 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/Models/MimeType.swift:5]: Ah I thought you had a slicker way to do it with just string literals, not using raw values.

File: Sources/UtilityBeltNetworking/HTTPClient+Combine.swift (hunk 6 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:119]: > What happens if someone already added this validator, not knowing it was included by default?

So does this matter actually?  The way the validators work, is it throws an error when the first one fails, so duplicating the `ensureMimeType` check will still ultimately fail the request as expected?

File: Sources/UtilityBeltNetworking/Models/ResponseValidator.swift (hunk 5 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/Models/ResponseValidator.swift:38]: Ya, this is odd.  We did not do any status code checking in the regular request builder, but in the non-combine request builder, we validated status codes on every request.  We probably need some consensus here whether this is something that we mandate or wether we require callers to pass it in. Any thoughts here?  

This also affects how we would handle the `defaultValidators` property - currently in this PR we don't allow overriding them, similar to how the combine request builder just did that validation in the base function (no way to override that as well).  I can see an argument for both...

**Comment by bdrelling** [comment]
@rhogan11 let me know when this is ready for re-review? Just not sure since I know this isn't high priority work for you.

Also, I'm out June 16th-18th this week. Let me know if you want to set up time tomorrow to go through anything outstanding so I don't leave ya hanging while I'm gone.

**Comment by rhogan11** [comment]
@bdrelling ah I was waiting on your response to some of my responses before making changes given that they are all connected by some higher level questions.  Perhaps we can chat quickly tomorrow to get some consensus on those - probably easier to talk about them than going back and forth on a PR.

File: Sources/UtilityBeltNetworking/HTTPClient+Combine.swift (hunk 6 above)
Comment (rhogan11) [reply Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:119]: This is also a little tricky because the `Validator` object is a struct that contains a closure.  Given closures cannot be compared, I'm not sure how we would check if a certain validation objects exist...

I'm still not too worried about this check though - the worst thing that happens is that duplicate successful validation happens on each request, which seems relatively trivial.  Is there a better way to structure validators to make them comparable?  Perhaps having a class structure instead of structs, so that we could have a subclass that can be compared (for instance a MimeTypeValidator that stores mime types and check against them?)

Comment (rhogan11) [reply Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:121]: @bdrelling let me know what you think we should do here when you have a sec, I think this is the only major outstanding work to be done on this PR

Comment (bdrelling) [reply Sources/UtilityBeltNetworking/HTTPClient+Combine.swift:121]: We discussed all of this on a Zoom call, resolving it.

**Comment by bdrelling** [comment]
We connecting on Zoom again and re-aligned, so I'm gonna ignore any previous comments (and you can too) and I'll just keep an eye out for re-review.

**Review by bdrelling — APPROVED** [review]
